In [1]:
dataset_path='IMDB Dataset.csv'

In [2]:
import pandas as pd
import tensorflow as tf
from tensorflow.keras.preprocessing.text import one_hot

In [3]:
tf.__version__  #tensorflow version

'2.15.0'

In [4]:
df=pd.read_csv(dataset_path)

In [5]:
df.head()

review sentiment
0  One of the other reviewers has mentioned that ...  positive
1  A wonderful little production. <br /><br />The...  positive
2  I thought this was a wonderful way to spend ti...  positive
3  Basically there's a family where a little boy ...  negative
4  Petter Mattei's "Love in the Time of Money" is...  positive

In [6]:
df=df.iloc[0:50,]  #slicng the df for few rows only
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     50 non-null     object
 1   sentiment  50 non-null     object
dtypes: object(2)
memory usage: 932.0+ bytes


In [7]:
#you can check for stopwords here and remove them 

In [8]:
#combining all reviews into a single list , this needs to be done to create a vocab
list_of_reviews=[]
for i, review in enumerate(df['review']):
    if len(review)==0:
        print(i)
    list_of_reviews.append(review)
    

In [9]:
len(list_of_reviews)

50

In [10]:
list_of_reviews


["One of the other reviewers has mentioned that after watching just 1 Oz episode you'll be hooked. They are right, as this is exactly what happened with me.<br /><br />The first thing that struck me about Oz was its brutality and unflinching scenes of violence, which set in right from the word GO. Trust me, this is not a show for the faint hearted or timid. This show pulls no punches with regards to drugs, sex or violence. Its is hardcore, in the classic use of the word.<br /><br />It is called OZ as that is the nickname given to the Oswald Maximum Security State Penitentary. It focuses mainly on Emerald City, an experimental section of the prison where all the cells have glass fronts and face inwards, so privacy is not high on the agenda. Em City is home to many..Aryans, Muslims, gangstas, Latinos, Christians, Italians, Irish and more....so scuffles, death stares, dodgy dealings and shady agreements are never far away.<br /><br />I would say the main appeal of the show is due to the f

In [11]:
#lets count the no of words in list of  sentences

total_words= sum(len(sentence.split()) for sentence in list_of_reviews)
total_words

10958

In [12]:
#means we require vocab size of more than 10000

voc_size=50000

## One hot representation

In [13]:
one_hot_repr=[one_hot(words,voc_size) for words in list_of_reviews]

In [14]:
print(one_hot_repr)

[[35083, 11937, 12393, 43581, 35168, 13050, 18923, 33969, 38419, 47304, 33159, 38945, 7263, 14122, 21589, 7392, 17210, 4914, 44555, 692, 38933, 37337, 23361, 48334, 7159, 29643, 15149, 7637, 34130, 34130, 12393, 4672, 33617, 33969, 29482, 7637, 47744, 7263, 47197, 21861, 16695, 39994, 43258, 14736, 11937, 12375, 43603, 11656, 9563, 692, 20586, 12393, 33731, 35059, 25184, 7637, 37337, 23361, 37504, 22719, 10221, 41623, 12393, 3656, 48152, 45120, 12193, 37337, 10221, 37226, 41697, 37120, 15149, 39771, 6860, 37627, 8774, 45120, 12375, 21861, 23361, 16113, 9563, 12393, 22662, 29626, 11937, 12393, 33731, 34130, 34130, 210, 23361, 14033, 7263, 38933, 33969, 23361, 12393, 47075, 3963, 6860, 12393, 3197, 39610, 44241, 33911, 24475, 210, 35709, 38105, 27057, 47959, 48780, 38048, 48901, 38939, 11937, 12393, 13148, 46698, 21742, 12393, 46751, 29053, 8290, 30968, 39994, 7404, 2508, 4200, 19989, 23361, 37504, 39887, 27057, 12393, 842, 10773, 48780, 23361, 42441, 6860, 17290, 8018, 14967, 25097, 209

In [15]:
sum(len(sublist) for sublist in one_hot_repr)  #total no of vocabs 
max(len(sublist) for sublist in one_hot_repr)  #max length of any sentence

708

### Word embedding  Representation

In [16]:
from tensorflow.keras.layers import Embedding
from tensorflow.keras.models import Sequential
from tensorflow.keras.preprocessing.sequence import pad_sequences
import numpy as np

In [17]:
max_review_length=max(len(sublist) for sublist in one_hot_repr)
embedded_docs=pad_sequences(one_hot_repr,padding='pre',maxlen=max_review_length)

In [18]:
embedded_docs[0]  #first sentence vectors

array([    0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,     0,     0,     0,     0,     0,     0,     0,
           0,     0,

In [19]:
#dimension assign, feature dimesnsions

dim= 300

In [20]:
model=Sequential()
model.add(Embedding(input_dim=voc_size,output_dim=dim,input_length=max_review_length))
model.compile('adam','mse')    #adam is an optimizer and mse- mean square error is loss function

In [21]:
model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 embedding (Embedding)       (None, 708, 300)          15000000  
                                                                 
Total params: 15000000 (57.22 MB)
Trainable params: 15000000 (57.22 MB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


#### Output Shape: (None, 708, 300)

The output shape of this layer is (None, 708, 300).
None indicates the batch size, which is not fixed and can vary.
708 is the sequence length, meaning each input example consists of 708 tokens (words).
300 is the size of the embedding vector, meaning each word is represented by a 300-dimensional vector.
Param #: 15000000

This is the number of parameters in the Embedding layer. It is calculated as vocab_size * embedding_dim.
In this case, it seems that the vocabulary size is 50,000 (as 50,000 * 300 = 15,000,000).

In [22]:
model.predict(embedded_docs[0])

23/23 [==============================] - 0s 1ms/step


array([[ 0.03816148, -0.02905076, -0.0408239 , ..., -0.01123405,
        -0.03197433, -0.03456888],
       [ 0.03816148, -0.02905076, -0.0408239 , ..., -0.01123405,
        -0.03197433, -0.03456888],
       [ 0.03816148, -0.02905076, -0.0408239 , ..., -0.01123405,
        -0.03197433, -0.03456888],
       ...,
       [ 0.0498361 , -0.02750099,  0.01848981, ..., -0.01878768,
         0.03334172,  0.00302321],
       [-0.03600078, -0.00909228,  0.00664064, ..., -0.01563757,
         0.03799535,  0.03904707],
       [ 0.03340801, -0.03873632,  0.01661631, ..., -0.01751403,
         0.02302254, -0.00244361]], dtype=float32)

In [23]:
#embedded_docs[0]

###  Storing vector to PG vector database

In [ ]:
# !pip install langchain_cohere

In [25]:
#!pip install langchain_postgres
#!pip install "psycopg[binary,pool]"

In [26]:
import psycopg
from langchain_core.documents import Document
from langchain_postgres import PGVector
from langchain_cohere import CohereEmbeddings
from typing import List,Dict


connection_string=f"postgresql+psycopg://postgres:admin@127.0.0.1:5432/learning"

In [27]:
collection_name="imdb_reviews"
embedding=CohereEmbeddings()

In [37]:
class EmbeddingDocument(Document):
    def __init__(self,page_content:str,metadata:dict=None,embed:List[float]=None):
        #intialize the paren't class intializer 
        super().__init__(page_content=page_content,metadata=metadata)
        #intialize the additional embedding attribute
        self.embed=embed

    


In [39]:
#docs=[EmbeddingDocuments(page_content=review,metadata={"id":i},embed=1) for i,review in enumerate(list_of_reviews)]

In [40]:
vector=model.predict(embedded_docs)

2/2 [==============================] - 0s 9ms/step


In [41]:
vectorstores=PGVector(collection_name=collection_name,
                      connection=connection_string,
                      embeddings=model,
                      use_jsonb=True)

In [42]:
#vectorstores.add_documents(docs)

In [43]:
def generate_embeddings(split_documents,onehot_embedding,model):
    try:
        
        Documents=[EmbeddingDocument(page_content=doc,metadata={"id":i},embed=model.predict(embedding)) for i,(doc,embedding) in enumerate(zip(split_documents,onehot_embedding))]
            
        return Documents
    except Exception as e:
        print(f"Error in generate_embeddings:{e}")
        
    

In [44]:
generate_embeddings(list_of_reviews[:2],embedded_docs[:2],model)

23/23 [==============================] - 0s 673us/step
Error in generate_embeddings:"EmbeddingDocument" object has no field "embed"


#### Code is working till connection pg vector , but unabel to creata subclass of document so that content and embedding can go in same metadata
